### PR TITLE
Feat: routing replay

### PR DESCRIPTION
**Pull request by DaniilSergeev17**

### CODE REVIEW

File: d9d/module/block/moe/layer.py [hunk 1]
```diff
@@ -64,6 +71,25 @@ def __init__(
 
         self.tokens_per_expert = nn.Buffer(torch.empty((num_grouped_experts,), dtype=torch.int64), persistent=False)
 
+        # Set by a RouterReplayRecorder while it is installed; used to capture this layer's expert selection.
+        self._replay_recorder: RouterReplayRecorder | None = None
+        self._replay_layer_id: int | None = None
```
Comment (mrapplexz) [inline d9d/module/block/moe/layer.py]: Shouldn't a MoELayer know nothing about its internal ID within the RouterReplayRecorder? We can change their coupling so RouterReplayRecoder#capture and other methods relying on ID accept a layer instance (self) and then RouterReplayRecorder maps it to its internal ID without exposing it to a MoELayer

File: d9d/module/block/moe/layer.py [hunk 2]
```diff
@@ -91,27 +117,36 @@ def reset_stats(self):
         """Resets the expert load balancing counters."""
         self.tokens_per_expert.zero_()
 
-    def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
+    def forward(self, hidden_states: torch.Tensor, replay_indices: torch.Tensor | None = None) -> torch.Tensor:
```
Comment (mrapplexz) [inline d9d/module/block/moe/layer.py:117]: I believe we have kind of a split-interface here:

1. The write operation uses the RouterReplayRecorder
2. The read operation uses an explicit argument called `replay_indices`

What do you think about making the RouterReplayRecorder the main control plane for the whole replaying stuff? 

Something like this:

1. The write operation would still use the RouterReplayRecorder
2. The read operation would use the RouterReplayRecorder too, so the end-user would call some method that sets replay indices on it (from the Task), and then the MoELayer would just access it and get replay indices related to it.

This looks better architecturally to me, however I believe it is not compatible with the current pipelining implementation and is blocked by the dataflow related refactor I'm working on.

File: d9d/module/block/moe/layer.py [hunk 3]
```diff
@@ -91,27 +117,36 @@ def reset_stats(self):
...
 
-        routing_result: RoutingResult = self.router(hidden_states)
+        routing_result: RoutingResult = self.router(hidden_states, replay_indices=replay_indices)
         expert_indices = routing_result.selected_expert_indices
         expert_scores = routing_result.selected_probabilities
         self._update_tokens_per_expert(expert_indices)
+
+        if self._replay_recorder is not None and replay_indices is None:
+            assert self._replay_layer_id is not None  # noqa: S101 -- always set together with the recorder
+            self._replay_recorder.capture(self._replay_layer_id, expert_indices.reshape(*old_shape[:-1], -1))
```
Comment (mrapplexz) [inline d9d/module/block/moe/layer.py]: I believe this would break if somebody calls the same MoELayer multiple times in a row - e.g., gradient accumulation and pipelining

File: d9d/module/block/moe/replay.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+from .layer import MoELayer
+
+
+class RouterReplayRecorder:
+    """Records the per-layer expert selection of every Mixture-of-Experts layer in a model.
+
+    Installing the recorder onto a model binds every `MoELayer` to this recorder and assigns it a stable
+    integer id (the order in which the layers appear in ``model.modules()``). While installed, each MoE layer that
+    runs a non-replay forward writes its selected expert indices into this recorder. After the forward pass,
+    `tape` assembles those per-layer selections into a single tensor that can be fed back as the
```
Comment (mrapplexz) [inline d9d/module/block/moe/replay.py]: > assembles those per-layer selections into a single tensor

I was thinking about edge cases here and came up with a parallel MoE layers case. Like when it's not just a sequential MoE->MoE->MoE model.

I believe packing everything into a single tensor assumes that MoE layers are sequential and eventually will break. Could we not pack states into a single tensor and use multiple tensors mapped by module names instead?

File: d9d/module/block/moe/layer.py (hunk 1 above)
Comment (DaniilSergeev17) [reply d9d/module/block/moe/layer.py]: Yes, I suppose, I broke MoELayer abstraction, will fix it

File: d9d/module/block/moe/layer.py (hunk 2 above)
Comment (DaniilSergeev17) [reply d9d/module/block/moe/layer.py:117]: Yes, this looks much better. But yes, now it's impossible to do it (because of current PP scheduling, which calls the same layer multiple times with different microbatches mixed). So, it waits the new pipelining implementation

File: d9d/module/block/moe/layer.py (hunk 3 above)
Comment (DaniilSergeev17) [reply d9d/module/block/moe/layer.py]: Yes, because of overwriting, we lose previous, e.g. microbatches results (only last microbatch stores in buffer). I propose to append results during capture (not overwrite) and while taping (.tape) to concatenate them along batch dimension

File: d9d/module/block/moe/replay.py (hunk 4 above)
Comment (DaniilSergeev17) [reply d9d/module/block/moe/replay.py]: Yes, it's a bug, will fix it with stacking by names (to be able to use parallel calls, not consecutive calls, and so on), not by num_layers with fixed ordering

It was: (batch_size, num_layers, seq_len, top_k)
Propose: {"layers.0.mlp": tensor(batch_size, seq_len, top_k), "layers.1.mlp": tensor(...), ...}

File: d9d/module/block/moe/replay.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+                pass supplied ``replay_indices`` and therefore skipped recording).
+        """
+        if not self._captures:
+            raise RuntimeError("No routing was recorded; run a forward pass while the recorder is installed.")
+        if self._captures.keys() != self._names.keys():
+            raise RuntimeError(
+                f"Expected all {len(self._names)} bound layers to record, but {len(self._captures)} did. "
+                "Did the forward pass supply replay_indices (which skips recording)?"
+            )
+        return {self._names[layer]: torch.cat(chunks, dim=0) for layer, chunks in self._captures.items()}
```
Comment (mrapplexz) [inline d9d/module/block/moe/replay.py]: what if we'll have microbatches w/ different seqlens?

Comment (mrapplexz) [reply d9d/module/block/moe/replay.py]: like some tricky gradient accumulation case with seq packing
